discovery_engine_core/web-api/src/models.rs

```diff
@@ -91,23 +113,19 @@ pub(crate) struct InteractionRequestBody {
 pub(crate) struct UserId(String);
 
 impl UserId {
-    fn new(value: &str) -> Result<Self, Error> {
-        let value = urlencoding::decode(value).map_err(Error::UserIdUtf8Conversion)?;
-
-        if value.trim().is_empty() {
-            Err(Error::UserIdEmpty)
-        } else if value.contains('\u{0000}') {
-            Err(Error::UserIdContainsNul)
-        } else {
-            Ok(Self(value.to_string()))
-        }
+    fn new(id: String) -> Result<Self, IdValidationError> {
+        let id = urlencoding::decode(&id).map_err(IdValidationError::InvalidUtf8)?;
+
+        validate_id_from_string(&*id)?;
```

Comment:
:art: no need for additional dereferencing, a `&Cow<str>` also fits for a `impl AsRef<str>`
```suggestion
        validate_id_from_string(&id)?;
```